driver-core/src/test/functional/com/mongodb/internal/connection/DefaultConnectionPoolTest.java

```diff
@@ -533,7 +533,11 @@ private static void useConcurrently(final DefaultConnectionPool pool, final int
         for (int i = 0; i < concurrentUsersCount; i++) {
             if ((checkoutSync && checkoutAsync) ? i % 2 == 0 : checkoutSync) {//check out synchronously and check in
                 tasks.add(executor.submit(() -> {
-                    while (!(timeout.hasExpired() || Thread.currentThread().isInterrupted())) {
+                    // TODO-CSOT this and below seem like they cannot be refactored into a typical "timeout.run"
```

Comment:
Note that the `hasExpired` method appears to be used exclusively for stopping execution. The only places where we don't stop execution by throwing are [here](https://github.com/mongodb/mongo-java-driver/pull/1349/files#diff-e80f234d8c415657233df477aad007b21149b6c258e633449ff3f776c0ba8b69R273), and for determining `isLastAttempt` (via the inline under `TimeoutContext#hasExpired`). If we can find a (prod) usage where a throw wouldn't improve things, I see no harm in keeping, but otherwise I think we should steer into forcing a throw.

An aside: I don't understand how `lastAttempt` works. We can't know if an attempt is the last attempt until the end of an attempt. If we check during the attempt (to add error labels), we could get "not last", not add the labels, but at the end of the attempt the timer has run out and actually it "is last", so we end... but without the requisite labels. We should check this - maybe the RetryState cleanup should be done sooner, to avoid possible bugs.

Regarding the TimeUnit, I don't think that we should fail fast (preemptively expire), since some ensuing operations (internal code, externals that take nanos) might complete within nanoseconds, and I see no reasonable way to enforce or consistently reason about what might come after.

Reply:
The above aside does seem to be a bug, which I have recorded as [a comment on the final PR](https://github.com/mongodb/mongo-java-driver/pull/1215/files#r1543542029).